core/src/main/scala/org/apache/spark/sql/SnappyParser.scala

```diff
@@ -348,39 +348,44 @@ class SnappyParser(session: SnappySession)
       (d, s.asInstanceOf[Option[Duration]]))
   }
 
-  private def extractGroupingSet(
+  private final def extractGroupingSet(
       child: LogicalPlan,
       aggregations: Seq[NamedExpression],
       groupByExprs: Seq[Expression],
       groupingSets: Seq[Seq[Expression]]): GroupingSets = {
     val keyMap = groupByExprs.zipWithIndex.toMap
-    val bitmasks: Seq[Int] = groupingSets.map(set => set.foldLeft(0)((bitmap, col) => {
+    val numExpressions = keyMap.size
+    val mask = (1 << numExpressions) - 1
+    val bitmasks: Seq[Int] = groupingSets.map(set => set.foldLeft(mask)((bitmap, col) => {
       require(keyMap.contains(col), s"$col doesn't show up in the GROUP BY list")
-      bitmap | 1 << keyMap(col)
+      bitmap & ~(1 << (numExpressions - 1 - keyMap(col)))
     }))
     GroupingSets(bitmasks, groupByExprs, child, aggregations)
   }
 
   protected final def groupingSetExpr: Rule1[Seq[Expression]] = rule {
-    '(' ~ ws ~ (expression + (',' ~ ws)) ~ ws ~ ')'  ~>  ((gs: Seq[Expression]) => gs) |
+    '(' ~ ws ~ (expression + (',' ~ ws)) ~ ')' ~ ws ~>  ((gs: Seq[Expression]) => gs) |
```

Comment:
The action ~> part in this looks redundant (pops and pushes the same) so can be removed.